Save disk space by only caching dev packages into the internal 'sources' directory

kit-ty-kate opened the pull request:
Fixes #5448 
Fixes #4056
~Queued on #6912~

The `.opam-switch/sources` directory currently stores the source of every installed packages in the following format:
* `<pkgname>`: are for pinned packages
* `<pkgname>.<version>`: are for the rest

In https://github.com/ocaml/opam/pull/2825 (opam 2.0.0~beta) the change from `packages.dev` to `sources` was introduced to support (as i understand it) non-pinned dev packages (e.g. `ocaml-variants.5.4.0+trunk`). This is a good change but this also had for side-effect to store the source of every packages regardless of whether it's a dev package or not, which i think isn't a good change given that it duplicates data that's already in the archive cache that could instead be extracted very easily (the cost of extraction vs. copy is negligible).
As a side note, extracting the sources from their tarball would be way faster on Windows that extraction + copy for the same reason detailed in #5741 

~As it currently stands this PR is a WIP. It compiles but still has many unsolved issues (in particular it breaks caching for non-pinned dev packages) and the reftests fail, but i'm opening it anyway to show the rough area of the code that needs to change and in case anyone wants to take over this work while i'm doing something else.~

Post-WIP edit: Further improvements can be made by extracting the tarball directly into the build directory and would also f.i.x https://github.com/ocaml/opam/issues/6693, but i think it is simpler to do that separately in a future PR.

kit-ty-kate commented:
As a demonstration, `opam clean --switch-cleanup --all` which primarily cleans up these unnecessary directories, just removed 20GB of disk space on my main machine.

kit-ty-kate commented:
xref live test results: https://github.com/ocaml/opam/pull/6906#issuecomment-4306603679

rjbou commented on src/state/opamSwitchAction.ml:
```diff
@@ -66,6 +66,7 @@ let create_empty_switch gt ?synopsis ?repos ?invariant switch =
     OpamFilename.mkdir (OpamPath.Switch.lib_dir root switch config);
     OpamFilename.mkdir (OpamPath.Switch.stublibs root switch config);
     OpamFilename.mkdir (OpamPath.Switch.toplevel root switch config);
+    OpamFilename.mkdir (OpamPath.Switch.sources_dir root switch);
```
Why do we need to always create that directory ?

kit-ty-kate replied:
mainly for one of the tests. I've replaced that by a commit adding the necessary mkdir

rjbou replied:
There is no need to change opam behaviour to comply a test. The test can be changed to no longer run an `ls` but another command that don't fail if the directory is not found. At the time, I used `ls` because `sources` directory is always created by opam when fetching sources. Another solution is to create the directory by hand `### mkdir OPAM/sources-dir/.opam-switch/sources/ ` at the beginning of the test. It doesn't change what is tested : we want to highlight what source directory is still present after the command.

rjbou replied:
I didn't see, you already made the mkdir

kit-ty-kate replied:
>  Another solution is to create the directory by hand ### mkdir OPAM/sources-dir/.opam-switch/sources/ 

sorry if i wasn't clear, that's what i meant by: `I've replaced that by a commit adding the necessary mkdir` (see new first commit from 2 hours ago)

rjbou replied:
indeed, i misread, you was clear enough

rjbou reviewed (changes requested):
On the idea, lgtm, it is a good enhancement for disk usage saving. On the implementation, i don't think it should be done that way. The `source_dir` function should keep returning the sources directory, and when we want to bypass that source directory to extract directly on the build dir, it should be done directly and explicitly.

rjbou commented on src/client/opamAction.ml:
```diff
@@ -269,8 +269,12 @@ let download_shared_source st url nvs =
         not (OpamPackage.Set.mem nv st.pinned &&
+             let dir = match OpamSwitchState.source_dir st nv with
+               | None ->
+                 assert false (* impossible given match on st.pinned above *)
```
I liked the previous version that used directly `OpamPath.x` when needed. It avoided the `assert false`.

kit-ty-kate replied:
done in the latest fixup commit

rjbou commented on src/client/opamSolution.ml:
```diff
@@ -645,21 +645,22 @@ let parallel_apply t
-      let source_dir = source_dir nv in
       (if OpamFilename.exists_dir source_dir
        then (if not is_inplace then
                OpamFilename.copy_dir ~src:source_dir ~dst:build_dir)
```
By not differentiating the cases, we end up with src = dst = build_dir if the package is not pinned

rjbou commented on tests/reftests/sources-directory-cleaning.test:
```diff
@@ -48,8 +48,9 @@ The following actions will be performed:
 -> installed waste.1
 Done.
+### : mkdir in case the sources directory hasn't been created
+### mkdir OPAM/sources-dir/.opam-switch/sources/
```
```suggestion

### : mkdir in case the sources directory hasn't been created
### mkdir OPAM/sources-dir/.opam-switch/sources/
### opam switch create sources-dir --empty
### : at install
### opam install waste
The following actions will be performed:
=== install 1 package
- install waste 1

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-> retrieved waste.1  (file://${BASEDIR}/waste.tgz)
-> installed waste.1
Done.
```

rjbou commented on src/state/opamSwitchState.mli:
```diff
@@ -152,8 +152,12 @@ val dev_packages: 'a switch_state -> package_set
-    on wether it's pinned). *)
-val source_dir: 'a switch_state -> package -> dirname
+    on wether it is pinned or simply a dev package).
+    Returns [None] if it is neither pinned or a dev package. *)
```
```suggestion
    Returns [None] if it is neither pinned or a dev package as defined
    in {!is_dev_package}.  *)
```

rjbou commented on src/state/opamSwitchState.mli:
```diff
@@ -152,8 +152,12 @@ val dev_packages: 'a switch_state -> package_set
+    Returns [None] if it is neither pinned or a dev package. *)
+val source_dir: 'a switch_state -> package -> dirname option
+
+(** Returns [source_dir] or [build_dir] otherwise. *)
```
```suggestion
(** Returns {!source_dir} or {!OpamPath.Switch.build_dir} otherwise. *)
```

rjbou commented on src/state/opamUpdate.ml:
```diff
@@ -707,7 +711,9 @@ let download_package_source_t st url nv_dirs =
 
 let download_shared_package_source st url nvs =
```
The documentation should be updated in the mli to highlight the new directory destinaation

kit-ty-kate replied:
not relevant anymore

rjbou commented on tests/reftests/dot-install.test:
```diff
@@ -128,7 +128,7 @@ ACTION                          prepare_package_source: dot.~dev at ${BASEDIR}/O
 -> removed   dot.~dev
 Installing dot.~dev.
-TRACK                           after install: 0 elements, 0 added, scanned in 0.000s
+TRACK                           after install: 3 elements, 0 added, scanned in 0.000s
```
How does it affects dirtrack added ? 
With full debug, the diff with master is
```diff
-@@ -120,17 +122,106 @@ added: [
+@@ -120,17 +122,103 @@ added: [
  ### unset OPAMDEBUGSECTIONS
  ### OPAMDEBUG=-5
  ### opam reinstall dot --dry-run
@@ -94,10 +94,6 @@
  ACTION                          download_package: dot.~dev
 PARALLEL                        Job 62929499 finished
 PARALLEL                        Starting job 798933370 (worker 1/15 -/1 -/1): build dot.~dev
-SYSTEM                          rmdir ${BASEDIR}/OPAM/inst/.opam-switch/build/dot.~dev
-SYSTEM                          rm ${BASEDIR}/OPAM/inst/.opam-switch/build/dot.~dev/dot.install
-SYSTEM                          rm ${BASEDIR}/OPAM/inst/.opam-switch/build/dot.~dev/a-file
-SYSTEM                          mkdir ${BASEDIR}/OPAM/inst/.opam-switch/build/dot.~dev
  ACTION                          prepare_package_source: dot.~dev at ${BASEDIR}/OPAM/inst/.opam-switch/build/dot.~dev
 PARALLEL                        Job 798933370 finished
 PARALLEL                        Starting job 743026333 (worker -/15 1/1 -/1): remove dot.~dev
@@ -107,11 +103,12 @@
  -> removed   dot.~dev
 PARALLEL                        Job 743026333 finished
 PARALLEL                        Starting job 699714656 (worker -/15 1/1 -/1): install dot.~dev
-FILE(.install)                  Cannot find ${BASEDIR}/OPAM/inst/.opam-switch/build/dot.~dev/dot.install
-TRACK                           before install: 0 elements scanned in 0.000s
+FILE(.install)                  Read ${BASEDIR}/OPAM/inst/.opam-switch/build/dot.~dev/dot.install in 0.000s
+FILE(.install)                  Wrote ${BASEDIR}/OPAM/inst/.opam-switch/install/dot.install atomically in 0.000s
+TRACK                           before install: 3 elements scanned in 0.000s
  Installing dot.~dev.
-[ERROR] tracked after {}
- TRACK                           after install: 0 elements, 0 added, scanned in 0.000s
+[ERROR] tracked after { share/dot/a-file:F:12fc204edeae5b57713c5ad7dcb97d39, share/dot:D, share:D }
+ TRACK                           after install: 3 elements, 0 added, scanned in 0.000s
  -> installed dot.~dev
 FILE(.config)                   Cannot find ${BASEDIR}/OPAM/inst/.opam-switch/config/dot.config
 PARALLEL                        Job 699714656 finished

kit-ty-kate committed:
Save disk space by only caching dev packages into the internal 'sources' directory

The 'sources' directory is still used for non-dev packages but only
temporarily during extraction and then moved to the 'build' directory
during build.

kit-ty-kate commented:
Thanks for the review. Turns out part of design had bad assumptions and extracting directly into `build` broke a lot of invariants that wasn't in the path that i tested for.

I redid the design and now the testsuite look good. The change is that we now use the `sources` directory as temporary directory that will be moved (they're pretty much guaranteed to be on the same partition so the cost of moving is none) to the `build` directory.

Depending on your personal preference i've pushed two commits. The first one changes the type of `OpamSwitchState.source_dir` to return whether or not it is meant to be permanent. However this value isn't used most of the time and costs an extra call to `OpamSwitchState.is_dev_package` so i also pushed a second commit to put back the old return type but add an extra `is_source_dir_temporary` for when the information is really necessary.

So depending on your preference, feel free to squash the second one or remove it, i'm happy with both solutions.

rjbou commented on src/state/opamSwitchState.ml:
```diff
@@ -808,6 +808,11 @@ let source_dir st nv =
+let is_source_dir_temporary st nv =
+  if OpamPackage.Set.mem nv st.pinned then false
+  else if is_dev_package st nv then false
+  else true
```
```suggestion
  not (OpamPackage.Set.mem nv st.pinned || is_dev_package)
```
??

rjbou commented on src/state/opamSwitchState.mli:
```diff
@@ -152,9 +152,14 @@ val dev_packages: 'a switch_state -> package_set
 val source_dir: 'a switch_state -> package -> dirname
 
+(** Returns [true] if the {!source_dir} of the package is meant to be a
+    temporary directory. *)
```
```suggestion
    temporary directory: packages that are not pinned or "dev package"
    as defined in {!is_dev_package}.  *)
```

rjbou commented on src/client/opamCommands.ml:
```diff
@@ -4355,7 +4355,7 @@ let clean cli =
                   (OpamPackage.Set.filter (OpamSwitchState.is_dev_package st)
                      st.installed))
-             |> List.map (OpamSwitchState.source_dir st)
+             |> List.map (fun nv -> OpamSwitchState.source_dir st nv)
```
This change is no longer needed in the last version

```suggestion
             |> List.map (OpamSwitchState.source_dir st)
```

rjbou reviewed (approved):
I like the second solution, it's more clear, and we don't have extra code (extra `snd`).

rjbou commented on master_changes.md:
```diff
@@ -31,6 +31,7 @@ users)
 
 ## Actions
   * Reorder the list of actions by increased priority [#6864 @kit-ty-kate - fix #6863]
+  * The internal `sources` directory now only serves to cache dev packages instead of every packages, saving disk space [#6440 @kit-ty-kate - fix #4056 #5448]
```
```suggestion
  * The internal `sources` directory now only serves to cache dev packages instead of every packages, saving disk space. It is removed after install. [#6440 @kit-ty-kate - fix #4056 #5448]
```

kit-ty-kate commented:
Ignoring the macOS failure (#7037)

kit-ty-kate commented:
Noteworthy: the `opam install lot-io-source lot-io-source2` benchmark show a 24% speedup :tada:
